src/openlcb/SimpleStack.hxx

```diff
@@ -358,40 +258,216 @@ protected:
-    Service service_{&executor_};
-    /// Abstract CAN bus in-memory.
-    CanHubFlow canHub0_{&service_};
-    /// NMRAnet interface for sending and receiving messages, formatting them
-    /// to the CAN bus port and maintaining the conversion flows, caches etc.
-    IfCan ifCan_{&executor_, &canHub0_, config_local_alias_cache_size(),
-        config_remote_alias_cache_size(), config_local_nodes_count()};
+    Service service_ {&executor_};
+    /// Pointer to the polymorphic implementation of the OpenLCB If.
+    std::unique_ptr<PhysicalIf> ifaceHolder_;
+    /// The OpenLCB interface object. Owned by ifaceHolder_;
+    If *iface_ {ifaceHolder_->iface()};
+    /// The datagram service bound to the interface object. Owned by
+    /// ifaceHolder_;
+    DatagramService *datagramService_ {ifaceHolder_->datagram_service()};
     /// Calls the config listeners with the configuration FD.
-    ConfigUpdateFlow configUpdateFlow_{&ifCan_};
+    ConfigUpdateFlow configUpdateFlow_ {iface()};
     /// The initialization flow takes care for node startup duties.
-    InitializeFlow initFlow_{&service_};
+    InitializeFlow initFlow_ {&service_};
     /// Dispatches event protocol requests to the event handlers.
-    EventService eventService_{&ifCan_};
+    EventService eventService_ {iface()};
     /// General flow for simple info requests.
-    SimpleInfoFlow infoFlow_{&ifCan_};
+    SimpleInfoFlow infoFlow_ {iface()};
 
-    CanDatagramService datagramService_{&ifCan_,
-        config_num_datagram_registry_entries(), config_num_datagram_clients()};
-    MemoryConfigHandler memoryConfigHandler_{
-        &datagramService_, nullptr, config_num_memory_spaces()};
+    MemoryConfigHandler memoryConfigHandler_ {
+        datagramService_, nullptr, config_num_memory_spaces()};
 
     /// All packets are forwarded to this hub in gridconnect format, if
     /// needed. Will be initialized upon first use.
     std::unique_ptr<HubFlow> gcHub_;
     /// Bridge between canHub_ and gcHub_. Lazily initialized.
     std::unique_ptr<GCAdapterBase> gcAdapter_;
-
     /// Stores and keeps ownership of optional components.
     std::vector<std::unique_ptr<Destructable>> additionalComponents_;
 };
 
-/// CAN-based stack with DefaultNode.
+/// SimpleStack with a CAN-bus based interface and IO functions for CAN-bus.
+class SimpleCanStackBase : public SimpleStackBase
+{
+public:
+    SimpleCanStackBase(const openlcb::NodeID node_id);
+
+    /// @returns the CanHubFlow to which this stack is talking to. This hub
+    /// flow usually has two members: the interface object from the software
+    /// stack and the hardware connection via which to connect to the physical
+    /// bus (which may be a device driver or a gridconnect protocol converter).
+    CanHubFlow *can_hub()
+    {
+        return &static_cast<CanPhysicalIf *>(ifaceHolder_.get())->canHub0_;
+    }
+
+    /// Adds a CAN bus port with synchronous driver API.
+    void add_can_port_blocking(const char *device)
```

Comment:
Should this be masked for Linux/Mach since it is not valid for esp32?

Reply:
This is only using POSIX APIs, so there should be no compilation problem on the esp32.
(There is no socketcan compatible fd-based driver in esp32 today, so it's not very useful but since it can compile it makes not a lot of sense to add conditional compilation around it.)

Reply:
Sure, just kinda thinking out loud a bit on it.